src/hotspot/os/posix/malloctrace/mallocTrace2.cpp

```diff
@@ -0,0 +1,2672 @@
+}
+
+bool MallocStatisticImpl::malloc_suspended() {
+  return _check_malloc_suspended && (pthread_getspecific(_malloc_suspended) != NULL);
+}
+
+void* MallocStatisticImpl::malloc_hook(size_t size, void* caller_address) {
+  void* result = _funcs->malloc(size);
+  uint64_t hash = ptr_hash(result);
+
+  if ((result != NULL) && should_track(hash) && !malloc_suspended()) {
+    address frames[MAX_FRAMES + FRAMES_TO_SKIP];
+    int nr_of_frames = capture_stack(frames, (address) malloc, (address) caller_address);
+
+    if (_track_free) {
+      record_allocation(result, hash, nr_of_frames, frames);
+    } else {
+      record_allocation_size(size, nr_of_frames, frames);
+    }
+  }
+
+  return result;
+}
+
+void* MallocStatisticImpl::calloc_hook(size_t elems, size_t size, void* caller_address) {
+  void* result = _funcs->calloc(elems, size);
+  uint64_t hash = ptr_hash(result);
+
+  if ((result != NULL) && should_track(hash) && !malloc_suspended()) {
+    address frames[MAX_FRAMES + FRAMES_TO_SKIP];
+    int nr_of_frames = capture_stack(frames, (address) calloc, (address) caller_address);
+
+    if (_track_free) {
+      record_allocation(result, hash, nr_of_frames, frames);
+    } else {
+      record_allocation_size(elems * size, nr_of_frames, frames);
+    }
+  }
+
+  return result;
+}
+
+void* MallocStatisticImpl::realloc_hook(void* ptr, size_t size, void* caller_address) {
+  size_t old_size = ptr != NULL ? _funcs->malloc_size(ptr) : 0;
+  uint64_t old_hash = ptr_hash(ptr);
+
+  // We have to speculate the realloc does not fail, since realloc itself frees
+  // the ptr potentially and another thread might get it from malloc and tries
+  // to add to the alloc hash map before we could remove it here.
+  StatEntry* freed_entry = NULL;
+
+  if (_track_free && (ptr != NULL) && should_track(old_hash)) {
+    freed_entry = record_free(ptr, old_hash, old_size);
+  }
+
+  void* result = _funcs->realloc(ptr, size);
+
+  if ((result == NULL) && (freed_entry != NULL) && (size > 0)) {
+    // We failed, but we already removed the freed memory, so we have to re-add it.
+    record_allocation(ptr, old_hash, freed_entry->nr_of_frames(), freed_entry->frames());
+
+    return NULL;
+  }
+
+  uint64_t hash = ptr_hash(result);
+
+  if ((result != NULL) && should_track(hash)) {
+    address frames[MAX_FRAMES + FRAMES_TO_SKIP];
+    int nr_of_frames = capture_stack(frames, (address) realloc, (address) caller_address);
+
+    if (_track_free) {
+      if (should_track(hash) && !malloc_suspended()) {
+        record_allocation(result, hash, nr_of_frames, frames);
+      }
+    } else if ((old_size < size) && should_track(hash) && !malloc_suspended()) {
+      // Track the additional allocate bytes. This is somewhat wrong, since
+      // we don't know the requested size of the original allocation and
+      // old_size might be greater.
+      record_allocation_size(size - old_size, nr_of_frames, frames);
+    }
+  }
```

Comment:
You can hoist the condition `should_track(hash) && !malloc_suspended()` up to the if statement in L1009
```
@@ -1006,15 +1006,13 @@ void* MallocStatisticImpl::realloc_hook(void* ptr, size_t size, void* caller_add
 
   uint64_t hash = ptr_hash(result);
 
-  if ((result != NULL) && should_track(hash)) {
+  if ((result != NULL) && should_track(hash) && !malloc_suspended()) {
     address frames[MAX_FRAMES + FRAMES_TO_SKIP];
     int nr_of_frames = capture_stack(frames, (address) realloc, (address) caller_address);
 
     if (_track_free) {
-      if (should_track(hash) && !malloc_suspended()) {
-        record_allocation(result, hash, nr_of_frames, frames);
-      }
-    } else if ((old_size < size) && should_track(hash) && !malloc_suspended()) {
+      record_allocation(result, hash, nr_of_frames, frames);
+    } else if (old_size < size) {
       // Track the additional allocate bytes. This is somewhat wrong, since
       // we don't know the requested size of the original allocation and
       // old_size might be greater.
```